src/simulations/collect.jl

```diff
@@ -0,0 +1,268 @@
+  dd[!, :id] = collect(keys(model.agents))
+  for fn in properties
+    dd[!, fn] = get_data.(values(model.agents), fn)
+  end
+  dd[!, :step] = fill(step, size(dd, 1))
+  return dd
+end
+
+"""
+    collect_model_data(model::ABM, properties::AbstractArray, step::Int)
+
+Collect model properties from functions or symbols provided in `properties`.
+"""
+function collect_model_data(model::ABM, properties::AbstractArray, step::Int)
+  dd = DataFrame()
+  for fn in properties
+    r =  get_data(model, fn)
+    dd[!, Symbol(fn)] = [r]
+  end
+  dd[!, :step] = fill(step, size(dd, 1))
+  return dd
+end
+
+"""
+    aggregate_data(df::AbstractDataFrame, aggregation_dict::Dict)
+  
+Aggregate `df` columns  with some function(s) specified in `aggregation_dict`.
+Each key in `aggregation_dict` is a column name (Symbol), and each value is
+an array of function to aggregate that column.
+
+Aggregation occurs per step.
+"""
+function aggregate_data(df::AbstractDataFrame, aggregation_dict::Dict)
+  all_keys = collect(keys(aggregation_dict))
+  dfnames = names(df)
+  available_keys = [k for k in all_keys if in(k, dfnames)]
+  length(available_keys) == 0 && return
+  v1 = aggregation_dict[available_keys[1]]
+  final_df = by(df, :step,  available_keys[1] => v1[1])
+  for v2 in v1[2:end]
+    dd = by(df, :step,  available_keys[1] => v2)
+    final_df = join(final_df, dd, on=:step)
+  end
+  for k in available_keys[2:end]
+    v = aggregation_dict[k]
+    for v2 in v
+      dd = by(df, :step,  k => v2)
+      final_df = join(final_df, dd, on=:step)
+    end
+  end
+
+  # rename columns
+  colnames = Array{Symbol}(undef, size(final_df, 2)) 
+  colnames[1] = :step
+  counter = 2
+  for k in available_keys
+    v = aggregation_dict[k]
+    for vv in v
+      colnames[counter] = Symbol(join([vv,"(", string(k), ")"], ""))
+      counter += 1
+    end
+  end
+  rename!(final_df, colnames)
+
+  return final_df
+end
+
+"used in _step!"
+function collect_agent_data!(df::DataFrame, model::ABM, agent_properties,  aggregation_dict, step::Int)
+  dft = collect_agent_data(model, agent_properties, step)
+  if isnothing(aggregation_dict)
+    df = vcat(df, dft)
+  else
+    dfa = aggregate_data(dft, aggregation_dict)
+    df = vcat(df, dfa)
+  end
+  return df
+end
+
+"used in _step!"
+function collect_model_data!(df::DataFrame, model::ABM, model_properties,  aggregation_dict, step::Int)
+  dft = collect_model_data(model, model_properties, step)
+  if isnothing(aggregation_dict)
+    df = vcat(df, dft)
+  else
+    dfa = aggregate_data(dft, aggregation_dict)
+    df = vcat(df, dfa)
+  end
+  return df
+end
+
+## step function
+###################
+
+until(ss, n::Int, model) = ss in 1:n
+until(ss, n::AbstractVector{Int}, model) = ss in n
+until(ss, n::Function, model) = !n(model)
+
+function _step!(model, agent_step!, model_step!, n, when; model_properties=nothing, aggregation_dict=nothing, agent_properties=nothing)
+  df_agent = DataFrame()
+  df_model = DataFrame()
+  ss = 0
```

Comment:
`until` needs `ss>0`, otherwise the loop will never run.

Reply:
There is `ss += 1` in `_step!`. Do you mean something else?

Reply:
No, I mean the initialisation. `ss=0`, then the call to `until(ss, n, model)`.  0 in 1:n will always return false.

Reply:
Oh yeah!